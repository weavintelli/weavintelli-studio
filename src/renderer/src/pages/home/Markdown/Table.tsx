import store from '@renderer/store'
import { messageBlocksSelectors } from '@renderer/store/messageBlock'
import { Tooltip } from 'antd'
import { Check, Copy } from 'lucide-react'
import React, { memo, useCallback, useState } from 'react'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'

interface Props {
  children: React.ReactNode
  node?: any
  blockId?: string
}

/**
 * 自定义 Markdown 表格组件，提供 copy 功能。
 */
const Table: React.FC<Props> = ({ children, node, blockId }) => {
  const { t } = useTranslation()
  const [copied, setCopied] = useState(false)

  const handleCopyTable = useCallback(() => {
    const tableMarkdown = extractTableMarkdown(blockId ?? '', node?.position)
    if (!tableMarkdown) return

    navigator.clipboard
      .writeText(tableMarkdown)
      .then(() => {
        setCopied(true)
        setTimeout(() => setCopied(false), 2000)
      })
      .catch((error) => {
        window.message?.error({ content: `${t('message.copy.failed')}: ${error}`, key: 'copy-table-error' })
      })
  }, [node, blockId, t])

  return (
    <TableWrapper className="table-wrapper">
      <table>{children}</table>
      <ToolbarWrapper className="table-toolbar">
        <Tooltip title={t('common.copy')} mouseEnterDelay={0.8}>
          <ToolButton role="button" aria-label={t('common.copy')} onClick={handleCopyTable}>
            {copied ? (
              <Check size={14} style={{ color: 'var(--color-primary)' }} data-testid="check-icon" />
            ) : (
              <Copy size={14} data-testid="copy-icon" />
            )}
          </ToolButton>
        </Tooltip>
      </ToolbarWrapper>
    </TableWrapper>
  )
}

/**
 * 从原始 Markdown 内容中提取表格源代码
 * @param blockId 消息块 ID
 * @param position 表格节点的位置信息
 * @returns 源代码
 */
export function extractTableMarkdown(blockId: string, position: any): string {
  if (!position || !blockId) return ''

  const block = messageBlocksSelectors.selectById(store.getState(), blockId)

  if (!block || !('content' in block) || typeof block.content !== 'string') return ''

  const { start, end } = position
  const lines = block.content.split('\n')

  // 提取表格对应的行（行号从1开始，数组索引从0开始）
  const tableLines = lines.slice(start.line - 1, end.line)
  return tableLines.join('\n').trim()
}

const TableWrapper = styled.div`
  position: relative;

  .table-toolbar {
    border-radius: 4px;
    opacity: 0;
    transition: opacity 0.2s ease;
    transform: translateZ(0);
    will-change: opacity;
  }
  &:hover {
    .table-toolbar {
      opacity: 1;
    }
  }
`

const ToolbarWrapper = styled.div`
  position: absolute;
  top: 8px;
  right: 8px;
  z-index: 10;
`

const ToolButton = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 4px;
  cursor: pointer;
  user-select: none;
  transition: all 0.2s ease;
  opacity: 1;
  color: var(--color-text-3);
  background-color: var(--color-background-mute);
  will-change: background-color, opacity;

  &:hover {
    background-color: var(--color-background-soft);
  }
`

export default memo(Table)
