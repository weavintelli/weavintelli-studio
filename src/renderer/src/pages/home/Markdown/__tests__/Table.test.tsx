import { act, render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, it, vi } from 'vitest'

import Table, { extractTableMarkdown } from '../Table'

const mocks = vi.hoisted(() => {
  return {
    store: {
      getState: vi.fn()
    },
    messageBlocksSelectors: {
      selectById: vi.fn()
    },
    windowMessage: {
      error: vi.fn()
    }
  }
})

// Mock dependencies
vi.mock('@renderer/store', () => ({
  __esModule: true,
  default: mocks.store
}))

vi.mock('@renderer/store/messageBlock', () => ({
  messageBlocksSelectors: mocks.messageBlocksSelectors
}))

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string) => key
  })
}))

vi.mock('antd', () => ({
  Tooltip: ({ children, title }: any) => (
    <div data-testid="tooltip" title={title}>
      {children}
    </div>
  )
}))

Object.assign(window, {
  message: mocks.windowMessage
})

describe('Table', () => {
  beforeAll(() => {
    vi.stubGlobal('jest', {
      advanceTimersByTime: vi.advanceTimersByTime.bind(vi)
    })
  })

  beforeEach(() => {
    vi.clearAllMocks()
    vi.useFakeTimers()
  })

  afterEach(() => {
    vi.restoreAllMocks()
    vi.runOnlyPendingTimers()
    vi.useRealTimers()
  })

  afterAll(() => {
    vi.unstubAllGlobals()
  })

  // https://testing-library.com/docs/user-event/clipboard/
  const user = userEvent.setup({
    advanceTimers: vi.advanceTimersByTime.bind(vi),
    writeToClipboard: true
  })

  // Test data factories
  const createMockBlock = (content: string = defaultTableContent) => ({
    id: 'test-block-1',
    content
  })

  const createTablePosition = (startLine = 1, endLine = 3) => ({
    start: { line: startLine },
    end: { line: endLine }
  })

  const defaultTableContent = `| Header 1 | Header 2 |
|----------|----------|
| Cell 1   | Cell 2   |`

  const defaultProps = {
    children: (
      <tbody>
        <tr>
          <td>Cell 1</td>
          <td>Cell 2</td>
        </tr>
      </tbody>
    ),
    blockId: 'test-block-1',
    node: { position: createTablePosition() }
  }

  const getCopyButton = () => screen.getByRole('button', { name: /common\.copy/i })
  const getCopyIcon = () => screen.getByTestId('copy-icon')
  const getCheckIcon = () => screen.getByTestId('check-icon')
  const queryCheckIcon = () => screen.queryByTestId('check-icon')
  const queryCopyIcon = () => screen.queryByTestId('copy-icon')

  describe('rendering', () => {
    it('should render table with children and toolbar', () => {
      render(<Table {...defaultProps} />)

      expect(screen.getByRole('table')).toBeInTheDocument()
      expect(screen.getByText('Cell 1')).toBeInTheDocument()
      expect(screen.getByText('Cell 2')).toBeInTheDocument()
      expect(screen.getByTestId('tooltip')).toBeInTheDocument()
    })

    it('should render with table-wrapper and table-toolbar classes', () => {
      const { container } = render(<Table {...defaultProps} />)

      expect(container.querySelector('.table-wrapper')).toBeInTheDocument()
      expect(container.querySelector('.table-toolbar')).toBeInTheDocument()
    })

    it('should render copy button with correct tooltip', () => {
      render(<Table {...defaultProps} />)

      const tooltip = screen.getByTestId('tooltip')
      expect(tooltip).toHaveAttribute('title', 'common.copy')
    })

    it('should match snapshot', () => {
      const { container } = render(<Table {...defaultProps} />)
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('extractTableMarkdown', () => {
    beforeEach(() => {
      mocks.store.getState.mockReturnValue({})
    })

    it('should extract table content from specified line range', () => {
      const block = createMockBlock()
      const position = createTablePosition(1, 3)
      mocks.messageBlocksSelectors.selectById.mockReturnValue(block)

      const result = extractTableMarkdown('test-block-1', position)

      expect(result).toBe(defaultTableContent)
      expect(mocks.messageBlocksSelectors.selectById).toHaveBeenCalledWith({}, 'test-block-1')
    })

    it('should handle line range extraction correctly', () => {
      const multiLineContent = `Line 0
| Header 1 | Header 2 |
|----------|----------|
| Cell 1   | Cell 2   |
Line 4`
      const block = createMockBlock(multiLineContent)
      const position = createTablePosition(2, 4) // Extract lines 2-4 (table part)
      mocks.messageBlocksSelectors.selectById.mockReturnValue(block)

      const result = extractTableMarkdown('test-block-1', position)

      expect(result).toBe(`| Header 1 | Header 2 |
|----------|----------|
| Cell 1   | Cell 2   |`)
    })

    it('should return empty string when blockId is empty', () => {
      const result = extractTableMarkdown('', createTablePosition())
      expect(result).toBe('')
      expect(mocks.messageBlocksSelectors.selectById).not.toHaveBeenCalled()
    })

    it('should return empty string when position is null', () => {
      const result = extractTableMarkdown('test-block-1', null)
      expect(result).toBe('')
      expect(mocks.messageBlocksSelectors.selectById).not.toHaveBeenCalled()
    })

    it('should return empty string when position is undefined', () => {
      const result = extractTableMarkdown('test-block-1', undefined)
      expect(result).toBe('')
      expect(mocks.messageBlocksSelectors.selectById).not.toHaveBeenCalled()
    })

    it('should return empty string when block does not exist', () => {
      mocks.messageBlocksSelectors.selectById.mockReturnValue(null)

      const result = extractTableMarkdown('non-existent-block', createTablePosition())

      expect(result).toBe('')
    })

    it('should return empty string when block has no content property', () => {
      const blockWithoutContent = { id: 'test-block-1' }
      mocks.messageBlocksSelectors.selectById.mockReturnValue(blockWithoutContent)

      const result = extractTableMarkdown('test-block-1', createTablePosition())

      expect(result).toBe('')
    })

    it('should return empty string when block content is not a string', () => {
      const blockWithInvalidContent = { id: 'test-block-1', content: 123 }
      mocks.messageBlocksSelectors.selectById.mockReturnValue(blockWithInvalidContent)

      const result = extractTableMarkdown('test-block-1', createTablePosition())

      expect(result).toBe('')
    })

    it('should handle boundary line numbers correctly', () => {
      const block = createMockBlock('Line 1\nLine 2\nLine 3')
      const position = createTablePosition(1, 3)
      mocks.messageBlocksSelectors.selectById.mockReturnValue(block)

      const result = extractTableMarkdown('test-block-1', position)

      expect(result).toBe('Line 1\nLine 2\nLine 3')
    })
  })

  describe('copy functionality', () => {
    beforeEach(() => {
      mocks.messageBlocksSelectors.selectById.mockReturnValue(createMockBlock())
    })

    it('should copy table content to clipboard on button click', async () => {
      render(<Table {...defaultProps} />)

      const copyButton = getCopyButton()
      await user.click(copyButton)

      await waitFor(() => {
        expect(getCheckIcon()).toBeInTheDocument()
        expect(queryCopyIcon()).not.toBeInTheDocument()
      })
    })

    it('should show check icon after successful copy', async () => {
      render(<Table {...defaultProps} />)

      // Initially shows copy icon
      expect(getCopyIcon()).toBeInTheDocument()

      const copyButton = getCopyButton()
      await user.click(copyButton)

      await waitFor(() => {
        expect(getCheckIcon()).toBeInTheDocument()
        expect(queryCopyIcon()).not.toBeInTheDocument()
      })
    })

    it('should reset to copy icon after 2 seconds', async () => {
      render(<Table {...defaultProps} />)

      const copyButton = getCopyButton()
      await user.click(copyButton)

      await waitFor(() => {
        expect(getCheckIcon()).toBeInTheDocument()
      })

      // Fast forward 2 seconds
      act(() => {
        vi.advanceTimersByTime(2000)
      })

      await waitFor(() => {
        expect(getCopyIcon()).toBeInTheDocument()
        expect(queryCheckIcon()).not.toBeInTheDocument()
      })
    })

    it('should not copy when extractTableMarkdown returns empty string', async () => {
      mocks.messageBlocksSelectors.selectById.mockReturnValue(null)

      render(<Table {...defaultProps} />)

      const copyButton = getCopyButton()
      await user.click(copyButton)

      await waitFor(() => {
        expect(getCopyIcon()).toBeInTheDocument()
        expect(queryCheckIcon()).not.toBeInTheDocument()
      })
    })
  })

  describe('edge cases', () => {
    it('should work without blockId', () => {
      const propsWithoutBlockId = { ...defaultProps, blockId: undefined }

      expect(() => render(<Table {...propsWithoutBlockId} />)).not.toThrow()

      const copyButton = getCopyButton()
      expect(copyButton).toBeInTheDocument()
    })

    it('should work without node position', () => {
      const propsWithoutPosition = { ...defaultProps, node: undefined }

      expect(() => render(<Table {...propsWithoutPosition} />)).not.toThrow()

      const copyButton = getCopyButton()
      expect(copyButton).toBeInTheDocument()
    })
  })
})
