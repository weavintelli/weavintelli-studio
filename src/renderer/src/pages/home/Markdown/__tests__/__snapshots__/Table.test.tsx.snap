// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > rendering > should match snapshot 1`] = `
.c0 {
  position: relative;
}

.c0 .table-toolbar {
  border-radius: 4px;
  opacity: 0;
  transition: opacity 0.2s ease;
  transform: translateZ(0);
  will-change: opacity;
}

.c0:hover .table-toolbar {
  opacity: 1;
}

.c1 {
  position: absolute;
  top: 8px;
  right: 8px;
  z-index: 10;
}

.c2 {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 4px;
  cursor: pointer;
  user-select: none;
  transition: all 0.2s ease;
  opacity: 1;
  color: var(--color-text-3);
  background-color: var(--color-background-mute);
  will-change: background-color,opacity;
}

.c2:hover {
  background-color: var(--color-background-soft);
}

<div
  class="c0 table-wrapper"
>
  <table>
    <tbody>
      <tr>
        <td>
          Cell 1
        </td>
        <td>
          Cell 2
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="c1 table-toolbar"
  >
    <div
      data-testid="tooltip"
      title="common.copy"
    >
      <div
        aria-label="common.copy"
        class="c2"
        role="button"
      >
        <svg
          class="lucide lucide-copy"
          data-testid="copy-icon"
          fill="none"
          height="14"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="14"
            rx="2"
            ry="2"
            width="14"
            x="8"
            y="8"
          />
          <path
            d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
